src/JustSaying/Messaging/Middleware/Handle/HandlerMiddlewareBuilder.cs

```diff
@@ -0,0 +1,139 @@
+        public HandlerMiddlewareBuilder(IHandlerResolver handlerResolver, IServiceResolver serviceResolver)
+        {
+            ServiceResolver = serviceResolver;
+            HandlerResolver = handlerResolver;
+            _middlewares = new List<Func<HandleMessageMiddleware>>();
+        }
+
+        /// <summary>
+        /// Adds a middleware of type <see cref="TMiddleware"/> to the pipeline which will be resolved from the
+        /// <see cref="IServiceResolver"/>. It will be resolved once when the pipeline is built, and cached
+        /// for the lifetime of the bus.
+        /// </summary>
+        /// <typeparam name="TMiddleware">The type of the middleware to add.</typeparam>
+        /// <returns>The current HandlerMiddlewareBuilder.</returns>
+        public HandlerMiddlewareBuilder Use<TMiddleware>() where TMiddleware : MiddlewareBase<HandleMessageContext, bool>
+        {
+            _middlewares.Add(() => ServiceResolver.ResolveService<TMiddleware>());
+            return this;
+        }
+
+        /// <summary>
+        /// Adds the provided middleware instance to the pipeline.
+        /// </summary>
+        /// <param name="middleware">An instance of a middleware to add to the pipeline.</param>
+        /// <returns>The current HandlerMiddlewareBuilder.</returns>
+        public HandlerMiddlewareBuilder Use(HandleMessageMiddleware middleware)
+        {
+            _middlewares.Add(() => middleware);
+            return this;
+        }
+
+        /// <summary>
+        /// Adds a middleware to the pipeline. The Func&lt;HandleMessageMiddleware&gt; will be called once
+        /// when the pipeline is built and cached for the lifetime of the bus.
+        /// </summary>
+        /// <param name="middlewareFactory">A <see cref="Func{HandleMessageMiddleware}"/> that produces an
+        /// instance of a middleware to use in the pipeline.</param>
+        /// <returns>The current HandlerMiddlewareBuilder.</returns>
+        public HandlerMiddlewareBuilder Use(Func<HandleMessageMiddleware> middlewareFactory)
+        {
+            _middlewares.Add(middlewareFactory);
+            return this;
+        }
+
+        /// <summary>
+        /// Adds a HandlerInvocationMiddleware{TMessage} to the pipeline. An <see cref="IHandlerAsync{TMessage}"/>
+        /// will be resolved from JustSaying's <see cref="IHandlerResolver"/> for each message and invoked.
+        ///
+        /// This is added automatically as the innermost handler to all pipelines, and doesn't need
+        /// to be called manually.
+        /// </summary>
+        /// <typeparam name="TMessage"></typeparam>
+        /// <returns>The current HandlerMiddlewareBuilder.</returns>
+        /// <exception cref="InvalidOperationException">
+        /// If a HandlerInvocationMiddleware already exists in this pipeline, it cannot be added again.
+        /// </exception>
+        public HandlerMiddlewareBuilder UseHandler<TMessage>() where TMessage : Message
```

Comment:
`internal` perhaps?

Reply:
This has been changed to internal. If in the future we want to open up the whole pipeline for customisation (so e.g. you could use a delegate as a handler instead of an `IHandlerAsync`) then we can re-think how we supply this pipeline through the API, but this should get us started.